src/api/server.rkt

```diff
@@ -53,75 +82,394 @@
+       ; Start a worker on a job. Unless the job-id is marked as finished then move to next state.
+       [(list 'start self command job-id)
+        (if (hash-has-key? completed-work job-id)
+            (place-channel-put self (list 'finished job-id (hash-ref completed-work job-id)))
+            (let ([worker (make-worker)])
+              ; Maybe this should be worker-id and we should pre allocate workers based on threads available.
+              (hash-set! workers job-id worker)
+              (when verbose
+                (eprintf "Starting worker [~a] on [~a].\n"
+                         job-id
+                         (test-name (herbie-command-test command))))
+              (place-channel-put worker (list 'apply self command job-id))))]
+       ; Job is finished notified the waiting list with the result of the job.
+       [(list 'finished job-id result)
+        (when verbose
+          (eprintf "Job ~a finished, saving result.\n" job-id))
+        ; Notifed job has been completed, save the result.
+        ; let GC collect worker 🤞.
+        (hash-set! completed-work job-id result)
+        (hash-remove! workers job-id)
+        (define maybe-wait-list (hash-ref waiting job-id #f))
+        (when maybe-wait-list
+          (when verbose
+            (eprintf "waiting job ~a completed\n" job-id))
+          (for ([waiting maybe-wait-list])
+            (place-channel-put waiting result))
+          (hash-remove! waiting job-id))]
+       ; Check if work is completed, returns the result or #f
+       [(list 'check job-id handler) (place-channel-put handler (hash-ref completed-work job-id #f))]
+       ; Pass a place-channel `handler` to the receptionist to be notified on when a job is complete.
+       [(list 'wait job-id handler)
+        ; first we add the handler to the wait list.
+        (if (false? (hash-ref waiting job-id #f))
+            (hash-set! waiting job-id (list handler))
+            (let ([wait-list (hash-ref waiting job-id)])
+              (hash-set! waiting job-id (append wait-list (list handler)))))
+        (define result (hash-ref completed-work job-id #f))
+        (when verbose
+          (eprintf "Waiting for job: ~a\n" job-id))
+        ; check if the job is completed oor not.
+        (unless (false? result)
+          ; we have a result to send.
+          (let ([maybe-wait-list (hash-ref waiting job-id #f)])
+            (unless (false? maybe-wait-list)
+              ; we have a waiting list to notify
+              (for ([waiting maybe-wait-list])
+                (place-channel-put waiting result))
+              (hash-remove! waiting job-id))))]))))
+
+(define receptionist #f)
 
-(define (start-job-server config global-demo global-output)
-  ;; Pass along local global values
-  ;; TODO can I pull these out of config or not need ot pass them along.
-  (set! *demo?* global-demo)
-  (set! *demo-output* global-output)
-  (thread-send *worker-thread* config))
+(define-syntax (place/context* stx)
```

Comment:
Copied from `thread-pool.rkt`, right?

Reply:
Ya, again copying what seems to work with out fully understanding what it's doing.